src/Utility/Hash.php

```diff
@@ -1146,6 +1147,224 @@ public static function mergeDiff(array $data, array $compare): array
         return $data;
     }
 
+    /**
+     * Validates data against provided schema.
+     *
+     * @param array $data Data to validate
+     * @param array $schema Schema to validate against
+     * @param bool $strict Whether all keys in $data must exist in $schema
+     * @return array
+     */
+    public static function validate(array $data, array $schema, bool $strict = true): array
+    {
+        $fields = $schema['fields'] ?? [];
```

Comment:
Only reason `fields` exists now is to allow for a `version`. If we don't have an outer array then I can't tell the different between an array of fields with a `version` field or the schema itself.

Reply:
Sounds good, better to keep a provision for more keys. I was wonder then if we could allow a global `required` key too then instead of having to specify it for each field. That could also allowing doing away with the `$strict` argument.

Reply:
We could. 

`$strict` basically allows pass-through data. `required` flag on the schema itself might feel like we're saying the default `required` setting for all the fields is that value. 

Do you think having a `strict` key in the schema would make more sense?

Reply:
> required flag on the schema itself might feel like we're saying the default required setting for all the fields is that value.

That's what I was suggesting. Guess I misunderstood what `$strict` did.

> Do you think having a strict key in the schema would make more sense?

Yes, I would prefer if the schema was self contained and we didn't have to rely to a special external flag to control how the validation works.

Reply:
`strict` and `required` are now supported at the schema level. The `strict` flag can be overwritten per array shape `array{}` and `required` can be overwritten per field.